### PR TITLE
feat: fifo queue

### PR DESCRIPTION
**Pull request by eladcon**
A queue but it is fifo and with `groupId` so theres order between messages in groups

```wing
bring fifoqueue;

let queue = new fifoqueue.FifoQueue();

queue.setConsumer(inflight (message: str) => {
  log("recieved message {message}");
});

test "will push to queue" {
  queue.push("a new message", groupId: "myGroup");
}
```

### CODE REVIEW

File: fifoqueue/fifo-queue.test.w [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+let counter = new cloud.Counter();
+
+queue.setConsumer(inflight (message: str) => {
+  log("start {message}");
+  util.sleep(10s);
+  counter.inc();
+  log("end {message}");
+});
+
+test "sequentially consume messages with the same group id " {
```
Comment (ekeren) [inline fifoqueue/fifo-queue.test.w:16]: Let's add a test for different groupId as well (that it is not blocking )

File: fifoqueue/fifo-queue.sim.w [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+pub class FifoQueue_sim impl api.IFifoQueue { 
+  queue: cloud.Queue;
+  counter: cloud.Counter;
+  
+  new(){
+    this.queue = new cloud.Queue();
+    this.counter = new cloud.Counter();
+  }
+
+  pub setConsumer(fn: inflight (str): void, options: api.SetConsumerOptions?) {
```
Comment (skyrpex) [inline fifoqueue/fifo-queue.sim.w:20]: What do you think about these renames?

`fn` to `handler`

File: fifoqueue/fifo-queue.sim.w [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    let counter = this.counter;
+    this.queue.setConsumer(inflight (m: str) => {
```
Comment (skyrpex) [inline fifoqueue/fifo-queue.sim.w:22]: `m` to `event`

File: fifoqueue/fifo-queue.sim.w [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+      let j = FifoQueueMessage.parseJson(m);
```
Comment (skyrpex) [inline fifoqueue/fifo-queue.sim.w:23]: `j` to `message`

File: fifoqueue/api.w [hunk 5]
```diff
@@ -0,0 +1,14 @@
+bring cloud;
+
+pub struct FifoQueueProps extends cloud.QueueProps{}
+
+pub struct PushOptions {
+  groupId: str;
+}
```
Comment (skyrpex) [inline fifoqueue/api.w:7]: I think we should also support a deduplicationId.

**Comment by skyrpex** [comment]
Maybe we should consider my fifo queue implementation too, it supports the deduplicationId but depends on DDB streams for the sim. See https://github.com/winglang/winglibs/pull/26/files#diff-18a9c2df571e1173005926ef3a6b1579bde0a9cdfa3a9c018420428408f45e35.

File: fifoqueue/fifo-queue.aws.w [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+      fifoQueue: true,
+      contentBasedDeduplication: true,
+    );
+    this.url = this.queue.url;
+    this.arn = this.queue.arn;
+  }
+
+  pub setConsumer(handler: inflight (str) : void, options: api.SetConsumerOptions?) {
+    let lambdaFn = new cloud.Function(inflight (event: str): void => {
+      let json: Json = unsafeCast(event);
```
Comment (eladb) [inline fifoqueue/fifo-queue.aws.w:34]: This is weird... @Chriscbr do we have an issue to track this?

File: fifoqueue/fifo-queue.aws.w [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+      let sqsEvent = SqsEvent.fromJson(event);
+      for message in sqsEvent.Records {
+        handler(message.body);
+      }
+    },
```
Comment (eladb) [inline fifoqueue/fifo-queue.aws.w:39]: something's off with the indentation

File: fifoqueue/api.w (hunk 5 above)
Comment (ekeren) [reply fifoqueue/api.w:7]: this is default AWS impl

**Review by ekeren — APPROVED** [review]

**Pull request merged**